### PR TITLE
Fix isTypeDerivedFrom to properly handle `{}` and intersections

### LINKED ISSUE #50844: `instanceof` does not narrow when the class is intersected with `&`
### 🔎 Search Terms

`instanceof`, `narrow`.

### 🕗 Version & Regression Information

- This is a crash
- This is the behavior in every version I tried (from `3.8.3` to to `4.9.0-dev.20220919`), and I reviewed the FAQ for entries about it.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&target=99&noPropertyAccessFromIndexSignature=true&exactOptionalPropertyTypes=true#code/JYOwLgpgTgZghgYwgAgJIgM5jiJB5EFAbwChlkB7QgCgEoAuZANwuABMSBfE0SWRFOiw4kAFQDuFZKXJhJdRi3ZcSCKlmShhuFAF5pnZHAxpM2HQRQAfU9rGSSq9WGQBhADbGMlgOpQqAObI+kSGxtLIhOLICrbm+ITIhgBkETAUFIxgUACuKNzAMDFa8SglIhAURR5evv4gAbTSZJpmFQB0VBB0yAD0vcgACv4ADtBgAJ7IAORd08hsFBAmIBQuEAAewFgANMgARhAIcDkYKAAG5TrnmitrkXBQ-uIQbAsU4iAqTmZunhjeQgAJWAAQAFi4QmETERIhBorEhKVLHt0plkNk8kkeEVqFckK07JVqv9ARAQeCwE0ZITSp0aE1+kNRuMprNCPNNtswCYcG9Lm1rrcHk8Pq93p8VEA)

### 💻 Code

```ts
interface InstanceOne {
  one(): void
}
interface InstanceTwo {
  two(): void
}
const instance = {} as InstanceOne | InstanceTwo

const ClassOneWrong = {} as { new (): InstanceOne } & { foo: true }
if (instance instanceof ClassOneWrong) {
  instance.one() // Property 'one' does not exist, because `instance` is not narrowed down
}

const ClassOneRight = {} as { new (): InstanceOne, foo: true }
if (instance instanceof ClassOneRight) {
  instance.one() // Property 'one' exists and `instance` is narrowed down
}
```

### 🙁 Actual behavior

`instanceof` does not narrow down the left side when the right side is a type using `&`.

### 🙂 Expected behavior

It should narrow it down to `InstanceOne`, and `instance.one()` should pass.

### LINKED ISSUE #51007: `instanceof Object` does not work with the `in` operator
# Bug Report

### 🔎 Search Terms

`instanceof Object` `in`

### 🕗 Version & Regression Information

- This changed between versions 4.9.0-dev.20220919 and 4.9.0-dev.20220920

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play?ts=4.9.0-dev.20220920#code/FAUwHgDg9gTgLgAgGYFcB2BjOBLKaHYDOAEgCoCyAMqQIYBGANiABRz1MBcC6A1mlAHc0ASi50oUJjXwBvYAgUIYIOChj4AhBraMQCAGT6EOpgTSE2mEFCQIA8nQBWILAaMByABZwAtg3dmxuwgANzAAL5AA)

### 💻 Code

```ts

export function isHTMLTable(table: unknown): boolean {
    return !!table && table instanceof Object && 'html' in table;
}
```

### 🙁 Actual behavior

Error: Type '{}' may represent a primitive value, which is not permitted as the right operand of the 'in' operator

### 🙂 Expected behavior

Should type check 

### Notes

Seems to be introduced by https://github.com/microsoft/TypeScript/pull/50666 which correctly introduces an error if `in` is used on non objects. The issue is that it does not handle testing for objects with `instanceof Object`

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #50844.
Fixes #51007.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -18759,16 +18759,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     // An object type S is considered to be derived from an object type T if
     // S is a union type and every constituent of S is derived from T,
     // T is a union type and S is derived from at least one constituent of T, or
-    // S is a type variable with a base constraint that is derived from T,
+    // S is an intersection type and some constituent of S is derived from T, or
+    // S is a type variable with a base constraint that is derived from T, or
+    // T is {} and S is an object-like type (ensuring {} is less derived than Object), or
```
Comment (Andarist) [inline src/compiler/checker.ts:18764]: q: what does "less derived" mean? how something can be more or less derived?

Comment (ahejlsberg) [reply src/compiler/checker.ts:18764]: It means that any object-like type, including `Object`, appears to be derived from `{}`. The effect of this addition is that `x instanceof Object` will narrow `{}` to `Object`, which in turn means that `x` qualifies as the right hand operand of an `in` check.

Comment (Andarist) [reply src/compiler/checker.ts:18764]: I'd probably reword this like this then:
```suggestion
    // T is {} and S is an object-like type (ensuring that object-likes, including Object, are derived from {}), or
```

Maybe I'm just missing the appropriate background and lingo but this explanation is more readable to me.

Comment (fatcerberus) [reply src/compiler/checker.ts:18764]: In the context of type theory, "less derived" is definitely a thing.  e.g. `Animal` is less derived than `Dog` which is less derived than `PitBull`, while `Cat` and `Dog` are equally derived, being sister types.

**Review by weswigham — APPROVED** [review]

**Pull request merged**